cogs/commands/quotes.py

```diff
@@ -0,0 +1,420 @@
+import logging
+import re
+from datetime import datetime
+from enum import Enum
+from typing import Optional
+
+from discord import AllowedMentions
+from discord.ext import commands
+from discord.ext.commands import Bot, Context, Converter
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.sql import func
+from sqlalchemy_utils import ScalarListException
+
+from models import db_session
+from models.quote import *
+from utils import (
+    get_database_user,
+    get_name_string,
+    is_compsoc_exec_in_guild,
+    user_is_irc_bot,
+)
+from utils.mentions import *
+
+LONG_HELP_TEXT = """
+Pull a random quote. Pull quotes by ID using "#ID", by author using "@username", or by topic by entering plain text
+"""
+SHORT_HELP_TEXT = """Record and manage quotes attributed to authors"""
+
+
+class QuoteError(Enum):
+    BAD_FORMAT = 1
+    NOT_PERMITTED = (2,)
+    NOT_FOUND = (3,)
+    OPTED_OUT = (4,)
+    DB_ERROR = (5,)
+    NO_OP = 6
+
+
+class QuoteException(Exception):
+    def __init__(
+        self,
+        err: QuoteError,
+        msg=None,
+    ):
+        self.message = (msg,)
+        self.err = err
+
+
+def is_id(string) -> bool:
+    if not isinstance(string, str):
+        return False
+    return re.fullmatch("#\d+", string) is not None
+
+
+"""check if mentioned user has opted out"""
+
+
+def user_opted_out(user: Mention, db_session=db_session) -> bool:
+    if user.is_id_type():
+        f = QuoteOptouts.user_id == user.id
+    else:
+        f = QuoteOptouts.user_string == user.string
+
+    q = db_session.query(QuoteOptouts).filter(f).one_or_none()
+
+    return q is not None
+
+
+def ctx_to_mention(ctx):
+    if user_is_irc_bot(ctx):
+        return MakeMention.string_mention(get_name_string(ctx))
+    else:
+        return MakeMention.id_mention(get_database_user(ctx.author).id)
+
+
+""" check if user has permissions for this quote """
+
+
+def has_quote_perms(is_exec, requester: Mention, quote: Quote):
+    if is_exec:
+        return True
+
+    if quote.author_type == MentionType.ID:
+        return requester.id == quote.author_id
+
+    return requester.string == quote.author_string
+
+
+def quote_str(q: Quote) -> Optional[str]:
+    if q is None:
+        return None
```

Comment:
Instead of checking for `None` here let the static typechecker do its work and don't bother guarding for it.